lib/src/values/config_number.cc

```diff
@@ -31,4 +37,32 @@ namespace hocon {
     bool config_number::operator!=(const config_number &other) const {
         return !(*this == other);
     }
+
+    int config_number::int_value_range_checked(std::string path) {
+        long l = long_value();
+        if (l < numeric_limits<int>::min() || l > numeric_limits<int>::max()) {
+            throw config_exception("Tried to get int from out of range value " + to_string(l));
+        }
+        return static_cast<int>(l);
+    }
+
+    unique_ptr<config_number> config_number::new_number(
+            shared_ptr<simple_config_origin> origin, int64_t value, std::string original_text) {
+        if (value >= numeric_limits<int>::min() && value <= numeric_limits<int>::max()) {
+            return unique_ptr<config_int>(new config_int(origin, static_cast<int>(value), original_text));
```

Comment:
The function new_number is taking origin by value, so it's already created a copy, and invalidating that copy will only affect other attempts to use it in the local scope (which there won't be, since you return after the move). By doing a move on shared_ptr, you're only moving the shared_ptr object, not the simple_config_origin it refers to.